src/client.rs

```diff
@@ -50,11 +50,11 @@ impl KeyStates {
 }
 
 fn send_client_message(msg: &ClientMessage, stream: &mut TcpStream) {
-    let data = serde_json::to_string(msg)
-        .expect("Failed to encode message");
-    stream.write(data.as_bytes())
-        .expect("Failed to send message to server");
-    stream.write(&[0])
+    let data = bincode::serialize(msg).expect("Failed to encode message");
+    let length = data.len() as u16;
+    stream.write(&length.to_be_bytes())
```

Comment:
Might be better to send the length and the data together as one message in case one succeeds and the other one fails

Reply:
Tcp doesn't have a concept of messages. It's stream oriented and bufffered. Besides we already split single "messages" on the server side if we don't send everything at once.